lib/logstash/inputs/beats.rb

```diff
@@ -99,104 +131,82 @@ def run(output_queue)
-      :value => ts.inspect, :exception => e.message)
+    @logger.debug("Beats input: stopped")
   end
 
-  private
-  def invoke(connection, &block)
-    @threadpool.post do
+  # This Method is called inside a new thread
+  def handle_new_connection(connection)
       begin
-        # If any errors occur in from the events the connection should be closed in the
-        # library ensure block and the exception will be handled here
-        connection.run do |map, identity_stream|
-          create_event(map, identity_stream, &block)
-        end
+        logger.debug? && logger.debug("Beats inputs: accepting a new connection",
+                                      :peer => connection.peer)
+
+        LogStash::Util.set_thread_name("[beats-input]>connection-#{connection.peer}")
+
+        connection_handler = LogStash::Inputs::BeatsSupport::ConnectionHandler.new(connection, self, @buffered_queue)
+        @connections_list[connection] = connection_handler
 
-        # When too many errors happen inside the circuit breaker it will throw
-        # this exception and start refusing connection. The bubbling of theses
-        # exceptions make sure that the lumberjack library will close the current
-        # connection which will force the client to reconnect and restransmit
-        # his payload.
+        # All the errors handling is done here
+        @circuit_breaker.execute { connection_handler.accept }
+      rescue Lumberjack::Beats::Connection::ConnectionClosed => e
+        logger.warn("Beats Input: Remote connection closed",
+                    :peer => connection.peer,
+                    :exception => e)
       rescue LogStash::CircuitBreaker::OpenBreaker,
-             LogStash::CircuitBreaker::HalfOpenBreaker => e
-        logger.warn("Beats input: The circuit breaker has detected a slowdown or stall in the pipeline, the input is closing the current connection and rejecting new connection until the pipeline recover.", :exception => e.class)
+        LogStash::CircuitBreaker::HalfOpenBreaker => e
+        logger.warn("Beats input: The circuit breaker has detected a slowdown or stall in the pipeline, the input is closing the current connection and rejecting new connection until the pipeline recover.",
+                    :exception => e.class)
       rescue => e # If we have a malformed packet we should handle that so the input doesn't crash completely.
-        @logger.error("Beats input: unhandled exception", :exception => e, :backtrace => e.backtrace)
+        @logger.error("Beats input: unhandled exception", 
+                      :exception => e,
+                      :backtrace => e.backtrace)
+      ensure
+        connection_handler.flush do |event|
+          # handle the basic event enrichment with tags
+          # since at that time we lose all the context
+          LogStash::Inputs::BeatsSupport::EventTransformCommon.new(self).transform(event)
```

Comment:
could we move the creation of the transformer to between line 197 and 198? otherwise we're creating 1 for each event, while it could be reused?